resource/cmake/CommonFunctions.cmake

```diff
@@ -0,0 +1,443 @@
+#  Licensee may redistribute copies of the software provided that all such    #
+#  copies are subject to this license.                                        #
+#  The software is provided "as is", with no warranty of any type, including  #
+#  any warranty for fitness for any purpose. RTI is under no obligation to    #
+#  maintain or support the software.  RTI shall not be liable for any         #
+#  incidental or consequential damages arising out of the use or inability to #
+#  use the software.                                                          #
+#                                                                             #
+###############################################################################
+
+function(copy_files)
+    set(BOOLEANS)
+    set(SINGLE_VALUE_ARGS CUSTOM_TARGET_NAME OUTPUT_DIR)
+    set(MULTI_VALUE_ARGS LIST_OF_FILES_RELATIVE_PATH)
+
+    cmake_parse_arguments(_COMMON_FUNCTIONS
+        "${BOOLEANS}"
+        "${SINGLE_VALUE_ARGS}"
+        "${MULTI_VALUE_ARGS}"
+        ${ARGN}
+    )
+
+    # Create the list of files to copy in the origin
+    foreach(FILE ${_COMMON_FUNCTIONS_LIST_OF_FILES_RELATIVE_PATH})
+        list(APPEND FILES_TO_COPY_ORIGIN "${CMAKE_CURRENT_SOURCE_DIR}/${FILE}")
+    endforeach(FILE ${_COMMON_FUNCTIONS_LIST_OF_FILES_RELATIVE_PATH})
+
+    # Create the list of files to copy in the output
+    foreach(FILE ${_COMMON_FUNCTIONS_LIST_OF_FILES_RELATIVE_PATH})
+        list(APPEND FILES_TO_COPY_OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/${FILE}")
+    endforeach(FILE ${_COMMON_FUNCTIONS_LIST_OF_FILES_RELATIVE_PATH})
+
+
+    # Create the custom command
+    add_custom_command(
+        # Indicate what files are generated if this command is executed
+        OUTPUT
+            ${FILES_TO_COPY_OUTPUT}
+        # Comment to show while building
+        COMMENT
+            "Copying ${_COMMON_FUNCTIONS_CUSTOM_TARGET_NAME} files to binary dir"
+        # Create the config directory if it doesn't exist
+        COMMAND
+            ${CMAKE_COMMAND} -E make_directory
+                "${_COMMON_FUNCTIONS_OUTPUT_DIR}"
+        # Copy the files (only if they are different)
+        COMMAND
+            ${CMAKE_COMMAND} -E copy_if_different
+                ${FILES_TO_COPY_ORIGIN}
+                "${_COMMON_FUNCTIONS_OUTPUT_DIR}"
+    )
+    # Create the custom target
+    add_custom_target(${_COMMON_FUNCTIONS_CUSTOM_TARGET_NAME}
+        # To execute this custom target (that will do nothing), we need first to
+        # generate these files. This triggers the custom command from the previous
+        # statement
+        DEPENDS
+            ${FILES_TO_COPY_OUTPUT}
+    )
+endfunction()
+
+
+###############################################################################
+
+function(generate_doc)
+    set(BOOLEANS DOXYGEN_ENABLED)
+    set(SINGLE_VALUE_ARGS PLUGIN_NAME CUSTOM_TARGET_NAME INPUT_DIR OUTPUT_DIR)
+    set(MULTI_VALUE_ARGS DEPENDENCIES)
```

Comment:
I'd prefer to keep meaningful names instead of short names. Are there any other reasons about shortening the variable names than saving some chars? (some conventions or something)

Reply:
The reason for shortening the variable is to make them clearer and easier to use. I'd say actually there is no reason to keep them so long and verbose. If you look at CMake's own function, there are hardly any argument with `_` in it. Here 4 out of 6 arguments do.

These are just too verbose, please update it.

Reply:
I don't think verbosity is a bad thing, probably because I'm not a CMake expert, but to me, that is something that provides more info about the variable (without looking at the documentation) and doesn't cost anything (neither time nor memory). 
In CMake functions, I think that they use whenever they need: for example, FILE uses several underscores: REMOVE_RECURSE , MAKE_DIRECTORY, TO_CMAKE_PATH or TO_NATIVE_PATH.

Reply:
```cmake
generate_doc(
  PLUGIN_NAME "GENERAL"
  CUSTOM_TARGET_NAME "GENERAL_DOC"
  INPUT_DIR "${CMAKE_CURRENT_SOURCE_DIR}"
  OUTPUT_DIR "${CMAKE_CURRENT_BINARY_DIR}/html"
  DEPENDENCIES ${DEPENDENCIES}
)
```

vs

```cmake
generate_doc(
  PLUGIN GENERAL
  TARGET GENERAL_DOC
  INPUT "${CMAKE_CURRENT_SOURCE_DIR}"
  OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/html"
  DEPENDS ${DEPENDENCIES}
)
```

The second is more concise and just as clear

Reply:
Btw, you picked a very specific example, where the argument is the name of the subcommand, which can hardly be considered a "generic parameter"

Reply:
I still not see the value of shortening some names (eg saying INPUT_DIR indicates that it is a directory, instead of INPUT). Anyway, I've done these changes to avoid delaying more the PR.